News feed with generalized widget customization component; persist user prefs in DB

acrellin opened the pull request:
[ch464]
[ch580]
[ch581]

This PR adds a news feed widget to the homepage and a generalized widget customization component that updates the user preferences on the back-end. In this example, changing the number of items to display updates the user preferences on the back-end, which then pushes a websocket message telling the front-end to re-fetch the newsfeed, and the back-end then supplies a new newsfeed with the new number of items specified in the user prefs in the DB. 

![news_feed](https://user-images.githubusercontent.com/7230285/74464009-9b196a00-4e47-11ea-8e16-d828fc8009e3.gif)

stefanv commented on skyportal/handlers/api/internal/profile.py:
```diff
@@ -27,4 +29,41 @@ def get(self):
+                                  'tokens': user_tokens,
+                                  'preferences': self.current_user.preferences})
+
+    def put(self):
```
We should probably define a schema for the configuration, so that these can be validated.

stefanv commented on skyportal/handlers/api/news_feed.py:
```diff
@@ -0,0 +1,59 @@
+        """
+        if (self.current_user.preferences and 'newsFeed' in self.current_user.preferences
+            and 'numItemsPerCategory' in self.current_user.preferences['newsFeed']):
+            n_items = int(self.current_user.preferences['newsFeed']['numItemsPerCategory'])
```
we need to put a cap on these.  `max(n_items, 20)` or something.

acrellin replied:
Done.

stefanv commented on skyportal/handlers/api/news_feed.py:
```diff
@@ -0,0 +1,59 @@
+            GroupSource.source_id).filter(GroupSource.group_id.in_(
+                [g.id for g in self.current_user.groups])
+            ))).order_by(desc(Source.modified)).limit(n_items).all()
+        photometry = Photometry.query.filter(Photometry.source_id.in_(DBSession.query(
```
This is repetitive and a bit verbose.  Can we refactor?

acrellin replied:
Not much -- I could write a function like `get_newest(model, order_by): ...` for the last two queries, but the first query has a different pattern. Do you still think that's worth it?

stefanv commented on skyportal/handlers/api/news_feed.py:
```diff
@@ -0,0 +1,59 @@
+                [g.id for g in self.current_user.groups])
+            ))).order_by(desc(Comment.created_at)).limit(n_items).all()
+
+        return self.success(
```
Let's think about how these are returned.  Something like a list with `{message: ..., type: ..., time: ...}` or some such may be better.  You can imagine that we'd want lots of different news sources eventually.  We may even eventually update SkyPortal to store such news events in a separate database (so we can include things like telescope alerts, etc.).

acrellin commented:
@stefanv I've addressed your previous review comments including updating the news feed item data structure to be more general and the same for each item, capping the number of news feed items, and adding a schema; I've added tests for the news feed and the widget customization component; and have pulled in the websocket message handler changes & updated accordingly.

acrellin commented on static/js/ducks/newsFeed.js:
```diff
@@ -0,0 +1,32 @@
+  }
+});
+
+export default function reducer(state=[], action) {
```
@stefanv those changes are not yet merged, so this is premature

stefanv replied:
That's right; didn't realize the others changes haven't made it in yet.

acrellin commented on skyportal/handlers/api/internal/profile.py:
```diff
@@ -27,4 +29,37 @@ def get(self):
+                                  'tokens': user_tokens,
+                                  'preferences': self.current_user.preferences})
+
+    def put(self):
```
Yes, this will be changed to patch in an upcoming PR -- for now, this is consistent with the rest of the app.

acrellin commented on skyportal/handlers/api/news_feed.py:
```diff
@@ -0,0 +1,57 @@
+              application/json:
+                schema: Error
+        """
+        if (self.current_user.preferences and 'newsFeed' in self.current_user.preferences
```
That's inconsistent with the rest of the DB schema -- all JSONB fields default to null, not an empty dictionary

acrellin commented on skyportal/handlers/api/news_feed.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        if (self.current_user.preferences and 'newsFeed' in self.current_user.preferences
+            and 'numItems' in self.current_user.preferences['newsFeed']):
+            n_items = min(int(self.current_user.preferences['newsFeed']['numItems']),
```
What is "the config file"?

stefanv replied:
Sorry, in our config reader.  We have a dictionary there that produces values for keys that do not exist.

acrellin commented on skyportal/handlers/api/news_feed.py:
```diff
@@ -0,0 +1,57 @@
+        news_feed_items.sort(key=lambda x: x['time'], reverse=True)
+        news_feed_items = news_feed_items[:n_items]
+
+        return self.success(data={'news_feed_items': news_feed_items})
```
I don't think so, but can change this to `newsFeedItems` if you'd like

stefanv replied:
As long as we're consistent---we should make sure we do the same thing everywhere.

acrellin commented on skyportal/schema.py:
```diff
@@ -46,6 +46,24 @@ class Error(Response):
+    newsFeed = fields.Nested(newsFeedPrefs)
+
+
+class UserPreferences(_Schema):
```
yes: https://github.com/skyportal/skyportal/pull/322/files#diff-be4099d3d54fe0c3ec39c6246f6fbcecR42

stefanv replied:
I don't think the docstrings are parsed and used for verification during the running of the app, is it?

acrellin replied:
I think I see what you mean -- do you want to import this schema in that handler and `load` the request parameters into it?

acrellin commented on static/js/components/NewsFeed.jsx:
```diff
@@ -0,0 +1,71 @@
+  const profile = useSelector((state) => state.profile);
+
+  const newsFeedPrefs = (
+    profile != null &&
```
Yes, but the components can try to mount even before there's a defaut-valued redux store, and the app breaks without this check. We do these kinds of things throughout the front-end, e.g. `if (!group || group.id === undefined) {...`

stefanv replied:
This should definitely not be the case.  The default state should be set by the time components are rendered.  Could you verify?

stefanv replied:
(It may be that our default state for that reducer is incorrect.)

acrellin replied:
Yes, I just verified this -- the default state is `state={ username: '', roles: [], acls: [], tokens: [], preferences: {} }` but when the component initially tries to mount, `profile` here is undefined, and the front-end crashes.

acrellin replied:
Oddly, the redux store sometimes fails to get properly initialized before the components first mount. I agree that the only check should be `Object.prototype.hasOwnProperty.call(profile.preferences, "newsFeed")`, which seems to work most of the time, so I'll change it back to that and investigate why the store isn't being initialized correctly at times.

acrellin replied:
The front-end tests error without this check -- perhaps the redux store isn't initialized correctly during testing. This is consistent with other components in the app, so I'm leaving this for now, and we can investigate what's going on during testing later.

acrellin commented on static/js/components/NewsFeed.jsx:
```diff
@@ -0,0 +1,71 @@
+          formValues={newsFeedPrefs}
+          stateBranchName="newsFeed"
+          title="News Feed Preferences"
+          onSubmit={ProfileActions.updateUserPreferences}
```
What? We _are_ passing `onSubmit` as a prop. What do you mean by "components are not supposed to be aware of redux state at all"? We frequently access the redux state with `useSelector`. Can you please clarify?

stefanv replied:
Yes, we previously had these sub-divided into components and containers.

acrellin replied:
These are all of the places we access the redux state from within components: 
```
static/js/components/CommentList.jsx:  const source = useSelector((state) => state.source);
static/js/components/CommentList.jsx:  const userProfile = useSelector((state) => state.profile);
static/js/components/CommentList.jsx:  const acls = useSelector((state) => state.profile.acls);
static/js/components/Footer.jsx:  const version = useSelector((state) => state.sysInfo.version);
static/js/components/Group.jsx:  const currentUser = useSelector((state) => state.profile);
static/js/components/Group.jsx:  const group = useSelector((state) => state.group);
static/js/components/GroupList.jsx:    groups = useSelector((state) => state.groups.latest);
static/js/components/Groups.jsx:  const roles = useSelector((state) => state.profile.roles);
static/js/components/HomePage.jsx:  const groups = useSelector((state) => state.groups.latest);
static/js/components/Logo.jsx:  const rotateLogo = useSelector((state) => state.logo.rotateLogo);
static/js/components/NewsFeed.jsx:  const { newsFeedItems } = useSelector((state) => state.newsFeed);
static/js/components/NewsFeed.jsx:  const profile = useSelector((state) => state.profile);
static/js/components/Plot.jsx:  const plots = useSelector((state) => state.plots);
static/js/components/Profile.jsx:  const profile = useSelector((state) => state.profile);
static/js/components/Profile.jsx:  const groups = useSelector((state) => state.groups.latest);
static/js/components/ProfileDropdown.jsx:  const profile = useSelector((state) => state.profile);
static/js/components/Source.jsx:  const source = useSelector((state) => state.source);
static/js/components/SourceList.jsx:  const sources = useSelector(
static/js/components/SourceList.jsx:  const sourceTableEmpty = useSelector(
static/js/components/SuperAdminGroupList.jsx:  const groups = useSelector((state) => state.groups.all);
static/js/components/UserInfo.jsx:  const users = useSelector((state) => state.users);
```

stefanv replied:
So, I guess at some point we got rid of the notion of containers vs components.  I don't have a problem with that.  Would be worth checking the redux docs to see their latest recommendation along with hooks.

acrellin replied:
Yes, that was the major refactor when we adopted React hooks. Our pattern follows the latest recommendations outlined by the `react-redux` team: https://react-redux.js.org/next/api/hooks#useselector

stefanv replied:
Dan Abramov:

> “ I don’t suggest splitting your components like this anymore. If you find it natural in your codebase, this pattern can be handy. But I’ve seen it enforced without any necessity and with almost dogmatic fervor far too many times. The main reason I found it useful was because it let me separate complex stateful logic from other aspects of the component. Hooks let me do the same thing without an arbitrary division.”

acrellin replied:
Agreed 100% -- so glad we moved away from the presentational/container component split and just use hooks now!

acrellin commented on static/js/components/ProfileDropdown.jsx:
```diff
@@ -39,6 +39,14 @@ const ProfileDropdown = () => {
 
+          <Link to="/sources" role="link">
+            <div className={styles.entry} onClick={collapseDropdown} role="menuitem" tabIndex="-1">
+              Sources
```
We have "groups" in the profile dropdown, too. When we add a sidebar, we'll move these links there. For now, this is the only way to get a page with just the source list.

acrellin commented on static/js/components/WidgetPrefsDialog.jsx:
```diff
@@ -0,0 +1,83 @@
+
+const WidgetPrefsDialog = ({ title, formValues, onSubmit, stateBranchName }) => {
+  const dispatch = useDispatch();
+  const [open, setOpen] = useState(false);
```
No, the material-ui docs recommend `open`

acrellin commented on skyportal/schema.py:
```diff
@@ -46,6 +46,24 @@ class Error(Response):
+    numItems = fields.String()
+
+
+class UserPrefs(_Schema):
```
To accurately represent the nested dictionary structure

stefanv replied:
I think the naming is just confusing.  I don't quite understand what this comment means, so would appreciate if you could clarify.

acrellin replied:
I nested these schemas in order to produce the nested dictionary structure of the request body. Here's an image of the rendered docs:

![image](https://user-images.githubusercontent.com/7230285/74485956-c4022500-4e70-11ea-9019-73875c109c92.png)

acrellin replied:
I could change the schema names to UpdateUserPreferencesRequestObject, and UserPrefs to UserPreferences, or something else, if that's clearer

acrellin replied:
I just pushed up that update: here's the resulting doc:

![image](https://user-images.githubusercontent.com/7230285/74486496-e9dbf980-4e71-11ea-8bd6-3cac7e44a47d.png)

acrellin commented on skyportal/tests/frontend/test_newsfeed.py:
```diff
@@ -0,0 +1,78 @@
+                           token=comment_token)
+        assert status == 200
+
+    driver.get(f'/become_user/{user.id}')
```
Without this, the new comments/sources don't appear in the browser session, because this user is a member of the group to which the new sources belong.

stefanv replied:
Is the source not posted as the current user?

acrellin replied:
No, it's posted with a token (`upload_data_token`), which is defined in `conftest` as belonging to the user associated with the fixture `user`.

acrellin replied:
```
@pytest.fixture()
def upload_data_token(user):
    token_id = create_token(permissions=['Upload data'],
                            created_by_id=user.id,
                            name=str(uuid.uuid4()))
    return token_id
```

acrellin commented on skyportal/tests/frontend/test_newsfeed.py:
```diff
@@ -0,0 +1,78 @@
+    for i in range(2):
+        driver.wait_for_xpath(f'//span[text()="New source {source_id_base}_{i}"]')
+        driver.wait_for_xpath(f'//span[contains(text(),"comment_text_{i} ({source_id_base}_{i})")]')
+    prefs_widget_button = driver.wait_for_xpath('//*[contains(@class,"MuiSvgIcon-root")]')
```
Done

acrellin commented on skyportal/handlers/api/news_feed.py:
```diff
@@ -0,0 +1,57 @@
+                schema: Error
+        """
+        if (self.current_user.preferences and 'newsFeed' in self.current_user.preferences
+            and 'numItems' in self.current_user.preferences['newsFeed']):
```
`self.current_user` is not a dictionary, so I'll change this to `preferences = self.current_user.preferences if self.current_user.preferences else {}`

acrellin commented on skyportal/handlers/api/internal/profile.py:
```diff
@@ -27,4 +29,37 @@ def get(self):
+        if not user_prefs:
+            user_prefs = data['preferences']
+        else:
+            user_prefs.update(data['preferences'])
```
Perfect, thanks for the tip

acrellin commented:
@stefanv I've pushed up changes to address your comments & suggestions. Ready for final review/merging.

acrellin commented:
@stefanv I believe I've addressed the latest round of comments - any other feedback/requests?

stefanv commented:
In it goes! :rocket:
